text/0732-cdk-cli-telemetry.md

```diff
@@ -0,0 +1,272 @@
+
+## Redacting Customer Content
+
+### Redacting CLI Commands
+
+We will redact free text input supplied by the user in the CLI command input.
+We know the available types of CDK CLI parameters and options, so this will be trivial to implement.
+Input values that are boolean or enums are not customer content and not redacted. All other strings are redacted.
+
+Take the following command:
+
+```bash
+> cdk bootstrap --bootstrap-bucket-name=MyBucket
+```
+
+Since `MyBucket` is free text, it will be redacted in the telemetry data sent to the service. This input will be processed into the following command object:
+
+```json
+{
+  "command": {
+    "path": ["cdk", "bootstrap"],
+    "parameters": {"bootstrap-bucket-name": "true" },
+  }
+}
+```
+
+We record that `bootstrap-bucket-name` is set, but not what the value is.
+
+Another example:
+
+```bash
+> cdk deploy MyStack
+```
+
+`MyStack` is free text, and also redacted. This command input turns into the following telemetry command:
+
+```json
+{
+  "command": {
+    "path": ["cdk", "deploy", "<redacted-stack-name>"],
+  }
+}
+```
+
+### Sanitizing Errors
+
+A core goal of the CDK CLI Telemetry Service is to provide a more efficient debugging experience for both the CDK core team and customers.
+This means that instead of back-and-forth on GitHub issues asking for more information — logs, stack trace, etc.
+
+When something goes wrong in the CDK CLI, we will report the error to the telemetry service.
+The error schema looks like this:
+
+```ts
+export type ErrorDetails = {
+  name: string;
+  message?: string; // anonymized error message
+  trace?: string; // anonymized stack trace
+  logs?: string; // anonymized stack logs
+}
+```
+
+The error name is an enum — currently either `ToolkitError`, `AuthorizationError`, `AssemblyError`, etc.
+
+The error message, trace, and logs could contain customer input.
+We have compiled the following list of possible customer inputs found in these strings and what we will do to redact those inputs:
+
+#### ARNs
+
+An ARN is a string that is prefixed by `arn:`. `arn:aws:s3:::my-bucket` will turn into `<redacted-arn>`.
+
+#### AccountIDs
+
+Anything that is a 12 digit number (that wasn’t already redacted within an ARN) will be redacted. `123456789012` will turn into `<redacted-account-id>`.
+
+#### File Paths
+
+Part of the file path (before node_modules) is customer content that we do not need to store.
+We will use a regex that includes the following to find and replace file path prefixes:
+
+```
+(file:/+)? -> matches optional file url prefix
+homedir()/process.cwd() -> users home directory or current working directory, replacing \ with /
```

Comment:
```suggestion
homedir()/process.cwd() -> users home directory or current working directory, replacing \ with /
```

Or just replace them with `$HOME` or `$PWD` respectively. Might be more readable.